toolchain/check/testdata/facet/no_prelude/combine.carbon

```diff
@@ -0,0 +1,138 @@
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// EXTRA-ARGS: --no-dump-sem-ir
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/facet/no_prelude/combine.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/facet/no_prelude/combine.carbon
+
+// --- core.carbon
+package Core;
+
+interface As(Dest:! type) {
+  fn Convert[self: Self]() -> Dest;
+}
+
+interface BitAnd {
+  fn Op[self: Self](other: Self) -> Self;
+}
+
+impl forall [T:! type] T as BitAnd {
+  fn Op[self: Self](other: Self) -> Self = "type.and";
+}
+
+// --- fail_combine_with_non_facet_type.carbon
+library "[[@TEST_NAME]]";
+
+import Core;
+
+interface A {}
+class C {}
+
+fn F() {
+  // CHECK:STDERR: fail_combine_with_non_facet_type.carbon:[[@LINE+8]]:23: error: BitAnd operator on types requires facet type on both sides [FacetTypeRequiredForTypeAndOperator]
+  // CHECK:STDERR:   ({} as C) as (C as (A & C));
+  // CHECK:STDERR:                       ^~~~~
+  // CHECK:STDERR:
+  // CHECK:STDERR: fail_combine_with_non_facet_type.carbon:[[@LINE+4]]:22: error: cannot evaluate type expression [TypeExprEvaluationFailure]
+  // CHECK:STDERR:   ({} as C) as (C as (A & C));
+  // CHECK:STDERR:                      ^~~~~~~
+  // CHECK:STDERR:
+  ({} as C) as (C as (A & C));
+  // CHECK:STDERR: fail_combine_with_non_facet_type.carbon:[[@LINE+8]]:23: error: BitAnd operator on types requires facet type on both sides [FacetTypeRequiredForTypeAndOperator]
+  // CHECK:STDERR:   ({} as C) as (C as (C & A));
+  // CHECK:STDERR:                       ^~~~~
+  // CHECK:STDERR:
+  // CHECK:STDERR: fail_combine_with_non_facet_type.carbon:[[@LINE+4]]:22: error: cannot evaluate type expression [TypeExprEvaluationFailure]
+  // CHECK:STDERR:   ({} as C) as (C as (C & A));
+  // CHECK:STDERR:                      ^~~~~~~
+  // CHECK:STDERR:
+  ({} as C) as (C as (C & A));
+  // CHECK:STDERR: fail_combine_with_non_facet_type.carbon:[[@LINE+12]]:23: error: BitAnd operator on types requires facet type on both sides [FacetTypeRequiredForTypeAndOperator]
+  // CHECK:STDERR:   ({} as C) as (C as (C & C));
+  // CHECK:STDERR:                       ^~~~~
+  // CHECK:STDERR:
+  // CHECK:STDERR: fail_combine_with_non_facet_type.carbon:[[@LINE+8]]:23: error: BitAnd operator on types requires facet type on both sides [FacetTypeRequiredForTypeAndOperator]
+  // CHECK:STDERR:   ({} as C) as (C as (C & C));
+  // CHECK:STDERR:                       ^~~~~
+  // CHECK:STDERR:
+  // CHECK:STDERR: fail_combine_with_non_facet_type.carbon:[[@LINE+4]]:22: error: cannot evaluate type expression [TypeExprEvaluationFailure]
+  // CHECK:STDERR:   ({} as C) as (C as (C & C));
+  // CHECK:STDERR:                      ^~~~~~~
+  // CHECK:STDERR:
```

Comment:
Should we try to only diagnose once in this case? i.e., only one `FacetTypeRequiredForTypeAndOperator`, no `TypeExprEvaluationFailure`

Reply:
Yeah, that appears to be done by returning error now.

Reply:
Note this still has a duplicate `FacetTypeRequiredForTypeAndOperator]`

Reply:
Yeah, this one does because it diagnoses both the lhs and the rhs, which I consider to be a good thing. We should have a better location so they are different though. I added a TODO comment.

Reply:
Also FWIW, using the `arg_id` instruction as the diagnostic location was what I wanted to do originally but they don't seem to have any location in them at all.

Reply:
SG, but this makes me want to comment more on the diagnostic itself.